# Mathmagical
## *Useful Math for Coding Interviews*

This document seeks to be a concise, simple explanation of some of the math you might come across in coding interviews.

### Sum of numbers 1 to n
The simplest way to do this is to just take the sum of every number between 1 and n (inclusive). It's readable, but it's also really inefficient.

In [1]:
# The naive solution...O(n) time. Yuck.
def slow_sum_to_n(n):
    sum = 0;
    for i in range(1, n + 1):
        sum += i
    return sum

print ("Sum 1 to 5:", slow_sum_to_n(5))
print ("Sum 1 to 10:", slow_sum_to_n(10))

Sum 1 to 5: 15
Sum 1 to 10: 55


If you think about it, there's a much better way to do this. Consider that these lists have the same sum:
>`[1, 2, 3, 4, 5]`

>`[3, 3, 3, 3, 3]`

**TLDR; the sum is n * the average value of the list, but it's not as simple as just dividing by 2.** It's tempting to just write math.ceil(x / 2), but that only works if n is odd.

If you have an even n (e.g. `[1, 2, 3, 4]`), the average is 2.5. This is because you take `n * avg_of_n`, where `avg_of_n = (n + 1) / 2`. 

Conveniently, this also works for an odd n (`[1, 2, 3, 4, 5]`)... `avg_of_n = (n + 1) / 2`. (which is the same as math.ceil() since n is odd). So in this case, the average is 3.

In [2]:
# The one-liner that runs in constant time (O(1))!
sum_to_n = lambda x: x * (x + 1) // 2

print ("Sum 1 to 5:", sum_to_n(5))
print ("Sum 1 to 10:", sum_to_n(10))

Sum 1 to 5: 15
Sum 1 to 10: 55


### Sum of Powers of 2
I love this one. If you want the sum of 2^0 to 2^n, you can do it without all that nasty loop business. To understand why, just look how these additions look in binary:

`
| n  | Binary |
|----|--------|
| 0  |  0001  |
| 1  |  0010  |
| 2  |  0100  |
| 3  |  1000  |
`

**The sum is 1111**

In [3]:
def sum_pow_two(n):
    return 2 ** (n + 1) - 1

print(sum_pow_two(0))
print(sum_pow_two(1))
print(sum_pow_two(2))
print(sum_pow_two(3))
print(sum_pow_two(4))
print(sum_pow_two(5))

1
3
7
15
31
63


### Number of Permutations
Say you're getting married and you're trying to plan out the seating arrangements. Because of the wonders of family drama and politics, it matters who sits next to who (Aunt Bertha can sit at the same table as Uncle Jim, but not right next to each other because Jim's allergic to Bertha's perfume). 

This is a perfect example for permutations, because in a permutation, the order matters. ABC != ACB.

Say each table holds 5 people, and you want to know all the possible ways your 50 guests could be seated at one of the tables. How many different ways can you do that?

For the first spot at your table, there are 50 possible guests. But for the second spot, you can't seat the same guest twice, so there are only 49 possibilities. Then 48, 47, and 46. Then whoops...you're out of chairs!

In [4]:
import math

def guest_permuations(table_size, guest_size):
    return math.factorial(guest_size) // math.factorial(guest_size - table_size)

# Generalized using mathematic notation P(n, k)
def P(n, k):
    # n! / (n-k)! isolates k * (k+1) * (k+2) * ... * n
    return math.factorial(n) // math.factorial(n - k)

print(P(7, 3))
print(P(50, 5))
print(P(6, 6)) # 0! = 1

210
254251200
720


#### Coding Note
Factorials are really awful in terms of time complexity: O(n) where n is the number. Especially when doing `n!/(n-k)!`, the entirety of the second call cancels out most of the first call. So while I wrote the above functions according to their mathematical notation, this would be a much more efficient solution:

In [5]:
# This runs in O(k) time.
def optimized_P(n, k):
    if (n < k):
        return 0;
    elif (k == 0):
        return 1 # Only option is empty set
    permutations = 1;
    for num in range(n, n - k, -1):
        permutations *= num
    return permutations

### Number of Combinations
Now let's pretend you live in a perfectly harmonious family where there's no drama. Anyone can sit next to anyone...magical. But that means that P(50, 5) is waaaay too many options! ABCDE == ABDEC == EDCBA, so we can use this to significantly reduce our choices.

Combinations are permutations where the order doesn't matter, denoted by C(n, k). To get a feel for how to solve our wedding problem, let's first look at a simpler example.

You have 3 cups and two balls. If you care which ball is in which cup, there are 6 possible combinations:

`
|       |    Balls (A, B, C)    |
|-------|---|---|---|---|---|---|
| Cup 1 | A | B | A | C | B | C |
| Cup 2 | B | A | C | A | C | B |
`


However, if we don't care which cup a ball is in, and only which balls are in cups, the number of options dwindles.

`
|     | Balls (A, B, C) |
|-----|-----|-----|-----|
| Cup |  A  |  A  |  B  |
| Cup |  B  |  C  |  C  |
`

#### How I think About it
If you have n items and you want to choose k, the first step is to compute P(n, k). However, this is the number of unique combinations of k elements of n.

Your sublist is k elements long. That means for any given list, there are k! permutations of it. Therefore, if you take the number of permutations (P(n, k)) and divide it by the number of ways k can be permutated, you get the number of combinations!

One more example...10 choose 3. There are 720 3-length permutations (`10 * 9 * 8`) of those 10 elements. Let's say the elements are letters: `[a, b, c, d, e, f, g, h, i, j]`

Of those 720 permutations, here are 6 of them:
* `[a, b, c]`
* `[a, c, b]`
* `[b, c, a]`
* `[b, a, c]`
* `[c, b, a]`
* `[c, a, b]`

Since we want combinations, all 6 (3! *ahem ahem*) of these are treated as the same permutation. This means for ever 6 permutations, we only care about one of them. Therefore the number of combinations is P(n, k) // k!...Or, written on its own: `n!/(n-k)!k!`

In [6]:
def C(n, k):
    numerator = math.factorial(n)
    denominator = math.factorial(n - k) * math.factorial(k)
    return numerator // denominator

# Optimized:
def optimized_C(n, k):
    return optimized_P(n, k) // math.factorial(k)

print(C(10, 3))

120


### Basic Probability
Ok, so I found a quarter on the sidewalk. Let's say I decide to start doing probability experiments with it. In my mind, there are three different types of probability I can measure:
* The probability of getting a specific order (e.g heads, tails, heads)
* The probability of getting exactly n heads (e.g. 2 heads out of 5 flips)
* The probability of getting *at least* n heads (e.g. 2 or more heads out of 5 flips)

In probability, you have an event with k possible outcomes. If you repeat that event n times, you will have k^n total possible outcomes. For example, flipping a coin 3 times could result in hhh, hht, hth, thh, htt, tht, tth, or ttt. This is 8 possible outcomes of the experiment, which is conveniently 2^3 (heads or tails 3 times).

If you want to know the probability of an event happening exactly in the order you want, your answer is 1/k^n because, of all the outcomes possible, only one meets your criteria.

If you don't care as much about the order of events, things get more interesting.

#### Probability of exactly 3 heads out of 5 flips.
You have a list, `[heads, heads, heads, tails, tails]`. The chances of getting exactly this outcome is 1/2^5, but if you don't care the order of the events, your chances are actually much higher. It may be tempting to think "Hey! I can just compute the number of permutations of this list, and divide it by my total number of outcomes to get the probability.", but this isn't quite the case. If you number the heads and tails, you start to see why.
>For our purposes, `[heads1, heads2, heads3, tails1, tails2] == [heads1, heads2, heads3, tails2, tails1]`

Hmm. Well, another way to think about it is heads1 can happen any of the 5 times you flip the coin, and heads2 can happen any time **except** when heads1 also happens. So instead of P(5, 5) = 120, you get P(5, 3) = 60 possible outcomes. This gets us closer because we're ignoring the tails, but it still assumes heads1 and heads2 are separate events and that we care about their order. Intuitively, you can also tell this is wrong because we only have 2^5 possible outcomes, so 60/32 gives us almost a 200% chance of our desired outcome...
>`[heads1, heads2, heads3, tails, tails] == [heads3, heads2, heads1, tails, tails]`

Since we don't care about the order...combinations seem like a great candidate. This would divide our 60 possible outcomes by 3! (the number of times we want heads), giving us 60/6 = 10. As our final step, we simply divide by the total number of outcomes to get our probability.

In [7]:
# Note that this only works if a given event only has two possible outcomes (heads, tails). 
# this wouldn't work for a dice, for example.
def probability_of_n_heads(heads, flips):
    desired_outcomes = optimized_C(flips, heads)
    total_outcomes = 2**flips
    percent = (desired_outcomes/total_outcomes) * 100
    return round(desired_outcomes/total_outcomes, 5)

print(probability_of_n_heads(0, 5))
print(probability_of_n_heads(1, 5) )
print(probability_of_n_heads(2, 5))
print(probability_of_n_heads(3, 5))
print(probability_of_n_heads(4, 5))
print(probability_of_n_heads(5, 5))

0.03125
0.15625
0.3125
0.3125
0.15625
0.03125


Woohoo! Just to really hit these percents home, this is what the outcomes look like:

`
Coin        1    2    3    4    5
outcome 1   h    h    h    h    h  (5 h)
        2   h    h    h    h    t  (4 h)
        3   h    h    h    t    h
        4   h    h    t    h    h
        5   h    t    h    h    h
        6   t    h    h    h    h
        7   h    h    h    t    t  (3 h)
        8   h    h    t    h    t
        9   h    t    h    h    t
       10   t    h    h    h    t
       11   h    h    t    t    h
       12   h    t    h    t    h
       13   t    h    h    t    h
       14   h    t    t    h    h
       15   t    h    t    h    h
       16   t    t    h    h    h
       17   t    t    t    h    h  (2 h)
       18   t    t    h    t    h
       19   t    h    t    t    h
       20   h    t    t    t    h
       21   t    t    h    h    t
       22   t    h    t    h    t
       23   h    t    t    h    t
       24   t    h    h    t    t
       25   h    t    h    t    t
       26   h    h    t    t    t 
       27   t    t    t    t    h  (1 h)
       28   t    t    t    h    t
       29   t    t    h    t    t
       30   t    h    t    t    t
       31   h    t    t    t    t
       32   t    t    t    t    t  (0 h)
`

#### Probability of *at least* 1 heads.
Generally, computing *at least* some amount is much harder, but there's a very convenient trick when you're looking for at least one: The probability of getting at least 1 heads is the same as 100% minus the probability of getting all heads. That's much easier to calculate, no? Here's a nice little one-liner.

In [8]:
def at_least_one_heads(flips):
    return 1 - 2**(-flips)

print(at_least_one_heads(0))
print(at_least_one_heads(1))
print(at_least_one_heads(2))
print(at_least_one_heads(3))
print(at_least_one_heads(4))
print(at_least_one_heads(10))

0
0.5
0.75
0.875
0.9375
0.9990234375


#### Probability of at least n
So this gets a little on the trickier side, and honestly probably not anything that will come up in a code review. But for a simple solution, it's useful to remember that you can either add the probabilities from n to the total number of events, **or** you can subtract the probabilities from 0 to n from 100%. Then just compute the smaller one. Kinda neat.

In [9]:
def at_least_n(n, flips):
    if n > flips / 2:
        return n_to_flips(n, flips)
    else:
        return zero_to_n(n, flips)
    
def n_to_flips(n, flips):
    total = 0;
    for i in range(n, flips + 1):
        total += probability_of_n_heads(i, flips)
    return round(total, 5)

def zero_to_n(n, flips):
    total = 0;
    for i in range(0, n):
        total += probability_of_n_heads(i, flips)
    return round(1.0 - total, 5)

print(at_least_n(0, 10))
print(at_least_n(1, 10))
print(at_least_n(2, 10))
print(at_least_n(3, 10))
print(at_least_n(4, 10))
print(at_least_n(5, 10))
print(at_least_n(6, 10))
print(at_least_n(7, 10))
print(at_least_n(8, 10))
print(at_least_n(9, 10))
print(at_least_n(10, 10))   

1.0
0.99902
0.98925
0.9453
0.82811
0.62303
0.37697
0.17189
0.0547
0.01075
0.00098


### Proof By Induction
This is a great way to prove your recursive function works. Conversely, if you can prove something by induction, that means there's a simple way to write it recurively (whether or not it's efficient is another matter).

There are three steps to prove something by induction... 
1. Prove the base case (P(b))
2. Assume your solution works for some value "n" (P(n))
3. Prove you can reach P(n+1) from P(n)

#### Example.
Prove that an n-element set contains 2^n subsets.

>##### Prove P(b).
>The base case is the empty set ({}). There's only one subset of the empty set, which is the empty set. This holds true because 2^0 = 1.

>##### Assume P(n)
>If we have an n element set, we assume there are 2^n subsets of it.

>##### Prove P(n + 1)
>This is the bread and butter of the problem. So let's say we have our 2^n subsets, then we add the n+1th element to the set. This would be all the subsets of n, plus all those subsets with the addition of n+1. For example:

>All possible subsets of `[a, b, c]`...
>* `[a, b, c]`
>* `[a, b]`
>* `[a, c]`
>* `[b, c]`
>* `[a]`
>* `[b]`
>* `[c]`
>* `[]`

>All possible subsets of `[a, b, c, d]` would be all of those in `[a, b, c]`, plus...
>* `[a, b, c, d]`
>* `[a, b, d]`
>* `[a, c, d]`
>* `[b, c, d]`
>* `[a, d]`
>* `[b, d]`
>* `[c, d]`
>* `[d]`

Therefore, if n had 2^n subsets, n+1 has 2 x 2^n, which is conveniently... 2^(n+1)

*QED.*